### PR TITLE
tracy: fix build on macos (fixes #223469)

### LINKED ISSUE #223469: Build failure: tracy
### Steps To Reproduce

1. build tracy on macOS (Apple Silicon)

### Build log

https://gist.github.com/maxammann/5a798d123e9df2d8b20f10a75296f878

### Notify maintainers

@mpickering
@nagisa

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"aarch64-darwin"`
 - host os: `Darwin 22.3.0, macOS 13.2.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.13.2`
 - channels(personal): `"darwin, home-manager-22.11.tar.gz"`
 - channels(max): `"darwin, home-manager-22.11.tar.gz"`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

### PR DESCRIPTION
**Pull request by maxammann**
###### Description of changes
The UniformTypeIdentifiers is required since some tracy version. This patch fixes the build on macOS

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

fixes https://github.com/NixOS/nixpkgs/issues/223469

### CODE REVIEW

**Comment by maxammann** [comment]
@nagisa this fixes the build. I'm not sure though why I need to access that framework through the darwin symbol, whereas AppKit does not require that.

**Comment by wegank** [comment]
`UniformTypeIdentifiers` is not available on macOS 10.12. Is there a way to build the package without this framework on `x86_64-darwin`?

**Comment by maxammann** [comment]
> `UniformTypeIdentifiers` is not available on macOS 10.12. Is there a way to build the package without this framework on `x86_64-darwin`?

So on macOS 10.12 the framework is not required: https://github.com/wolfpld/tracy/blob/33a640f848c77c6297b0f04dc6567d59eaed321b/nfd/nfd_cocoa.m#L12-L29

Can we somehow only add this framework for SDKs < 11.0? Should I just assert on the architecture of darwin?

**Comment by wegank** [comment]
You can do something like `lib.versionAtLeast stdenv.hostPlatform.darwinMinVersion "11"`.

**Comment by maxammann** [comment]
Oke, so this should work now, both on x86 and on latest Apple Silicon.

**Comment by wegank** [comment]
```
ld: framework not found UniformTypeIdentifiers
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```

You might need to remove `-framework UniformTypeIdentifiers` somewhere in the code, conditionally.

**Comment by maxammann** [comment]
I added a patch :) let's try again!

**Comment by wegank** [comment]
@ofborg build tracy tracy.passthru.tests

**Comment by wegank** [comment]
> Should the linux tests succeed? I did not really touch Linux here

They should: https://hydra.nixos.org/job/nixpkgs/trunk/tracy.x86_64-linux

**Comment by wegank** [comment]
Looks like `LEGACY=1` is applied in AUR: https://aur.archlinux.org/cgit/aur.git/commit/?h=tracy&id=a6bd906c2c7aee86bd5d0022c0c7cf2fdf7e4228

**Comment by maxammann** [comment]
Alright I blindly changed what was done on the AUR.

I dont have a linux machine with desktop available right now. So if somebody could test this, that would be great!

Else I can test in 2 months :)

File: pkgs/development/tools/tracy/default.nix [hunk 1]
```diff
@@ -1,25 +1,29 @@
...
+  patches = [ ]
+    ++ lib.optionals (stdenv.isDarwin && !(lib.versionAtLeast stdenv.hostPlatform.darwinMinVersion "11")) [ ./0001-remove-unifiedtypeidentifiers-framework ];
+
   nativeBuildInputs = [ pkg-config ];
 
   buildInputs = [ glfw capstone ]
     ++ lib.optionals stdenv.isDarwin [ Carbon AppKit freetype ]
-    ++ lib.optionals stdenv.isLinux [ gtk3 tbb dbus ];
+    ++ lib.optionals (stdenv.isDarwin && lib.versionAtLeast stdenv.hostPlatform.darwinMinVersion "11") [ darwin.apple_sdk.frameworks.UniformTypeIdentifiers ]
+    ++ lib.optionals stdenv.isLinux [ tbb dbus pkgs.hicolor-icon-theme freetype ];
```
Comment (wegank) [inline pkgs/development/tools/tracy/default.nix:26]: `pkgs.hicolor-icon-theme` doesn't look right. Anyway, I can fix this for you tomorrow or next Monday.

Comment (maxammann) [reply pkgs/development/tools/tracy/default.nix]: Ah yeah, pkgs is not there :)

Thanks! Let's see if it works and if not, that would be great!

Comment (maxammann) [reply pkgs/development/tools/tracy/default.nix]: I upgraded directly to 0.9.1 so the patch is compatible.

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 223835` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tracy</li>
  </ul>
</details>

**Pull request merged**